### PR TITLE
Enterprise SONiC "fips" resource module changes

### PR DESCRIPTION
**Pull request by balasubramaniam-k**
SUMMARY

1. Addition of Enterprise SONiC 'fips' resource module:
    Data definitions in 'argspec'.
    Implementation of 'facts' to gather facts for the module.
    Implementation of 'config' to apply configurations for merged and deleted states.
2. Definition of regression test cases to verify correct functionality of fips resource module.

ISSUE TYPE

New Module Pull Request

COMPONENT NAME

sonic_fips

DEPENDANT PR:

https://github.com/ansible-network/resource_module_models/pull/244

##### Checklist:

- [x] I have performed a self-review of my own code to ensure there are no formatting, linting, or security issues
- [x] I have verified that new and existing unit tests pass locally with my changes
- [x] I have not allowed coverage numbers to degenerate
- [x] I have maintained at least 90% code coverage
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature 

ADDITIONAL INFORMATION

Test Results
1. Regression test HTML report:
[fips_regression_report.pdf](https://github.com/ansible-collections/dellemc.enterprise_sonic/files/13860143/fips_regression_report.pdf)

2. Regression test detailed log:
[fips_regression_terminal_output.txt](https://github.com/ansible-collections/dellemc.enterprise_sonic/files/13860169/fips_regression_terminal_output.txt)

3. Gathering of facts:

[fips_get_facts_output.txt](https://github.com/ansible-collections/dellemc.enterprise_sonic/files/13860179/fips_get_facts_output.txt)

Regression test HTML report:
02-02-2024:
[fips_regression_report_02022024.pdf](https://github.com/ansible-collections/dellemc.enterprise_sonic/files/14139244/fips_regression_report_02022024.pdf)

### CODE REVIEW

**Comment by kerry-meyer** [comment]
Thank you for providing the change set to provide functionality for the FIPS resource module.

To proceed with PR review, please fix the sanity test errors. (See the "Details" link for the specific errors being flagged.)

**Comment by balasubramaniam-k** [comment]
> Thank you for providing the change set to provide functionality for the FIPS resource module.
> 
> To proceed with PR review, please fix the sanity test errors. (See the "Details" link for the specific errors being flagged.)

I have fixed the errors raised as part of sanity test.

File: plugins/module_utils/network/sonic/config/fips/fips.py [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+                  of the provided objects
+        """
+        if not want:
+            commands = have
+        else:
+            diff = get_diff(want, have)
+            commands = diff
+        requests = []
+
+        if have['enable'] is True:
```
Comment (xhan-dell) [inline plugins/module_utils/network/sonic/config/fips/fips.py:188]: deletes on booleans, strings, numbers, etc. should check want and have config have matching values before deleting/resetting the value

**Review by xhan-dell — CHANGES_REQUESTED** [review]
just some small things!

File: tests/unit/modules/network/sonic/fixtures/sonic_fips.yaml [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+      response:
+        code: 200
+        value:
+          openconfig-fips:config:
+            fips-mode: false
+  config_requests:
+    - path: "data/openconfig-fips:fips/config/fips-mode"
+      method: "patch"
+      data:
+        openconfig-fips:fips-mode: true
```
[the remark this answers is not in the document]
Comment (balasubramaniam-k) [reply tests/unit/modules/network/sonic/fixtures/sonic_fips.yaml]: Updated the test case accordingly

File: plugins/module_utils/network/sonic/config/fips/fips.py (hunk 1 above)
Comment (balasubramaniam-k) [reply plugins/module_utils/network/sonic/config/fips/fips.py]: Addressed.

File: plugins/module_utils/network/sonic/config/fips/fips.py [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+
+        :rtype: A list
+        :returns: the commands necessary to remove the current configuration
+                  of the provided objects
+        """
+        if not want:
+            commands = have
+        else:
+            diff = get_diff(want, have)
+            commands = diff
```
Comment (xhan-dell) [inline plugins/module_utils/network/sonic/config/fips/fips.py:185]:
```suggestion
            diff = get_diff(want, have)
            commands = get_diff(want, diff)
```
this change is made in context of the already committed change in the code following highlighted section (specifically the line checking if `enabled` is in commands). 
this sets commands to what is the same between want and have. Then the if statement will only trigger if both want and have enabled are true.

File: plugins/module_utils/network/sonic/config/fips/fips.py (hunk 1 above)
Comment (xhan-dell) [reply plugins/module_utils/network/sonic/config/fips/fips.py]: resolved with next comment's suggestion.

**Review by xhan-dell — APPROVED** [review]

**Pull request merged**